cunumeric/module.py

```diff
@@ -1220,6 +1220,27 @@ def __init__(self, ndim, shape, dtype):
         self.dtype = dtype
 
 
+def check_list_depth(active_arr, leaf_map):
```

Comment:
This function is still convoluted and makes me struggle to understand what it does. I'd write it in this way (ignoring the `leaf_map` stuff that I'm still trying to understand):

```
def check_list_depth(arr, prefix):
    if not isinstance(arr, list):
        return 0
    elif len(arr) == 0:
        raise ValueError(f"List at {prefix} cannot be empty")

    depths = [
        check_list_depth(each, (*prefix, idx))
        for idx, each in enumerate(arr)
    ]
    if len(set(depths)) != len(depths):
        # If we're here elements don't have the same depth
        first_depth = 1
        for other_depth in depths[1:]:
            if other_depth != first_depth:
                raise ValueError(
                    "List depths are mismatched. First element was at depth "
                    f"{depth}, but there is an element at depth {other_depth} "
                )

    return depths[0] + 1
```

Reply:
I do similar things in the first version I had. Tried to traverse graph only once, which makes the code more complicated. 
Will go back to this because the current implementation includes comparison of the indicies a lot.